### PR TITLE
Fix loss of images when merging patches against pre-existing create/patch deltas

### PR DESCRIPTION
**Pull request by nirvn**
The mysterious image loss issue has been resolved! 

Long story short, if a photo was changed/added in a follow up edit after a feature was added or already edited once _prior to pushing_, the photos would go missing as the logic to merge new patches into pre-existing create and patch deltas were not adding the necessary file checksums.

The PR fixes that. Woupidou.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 97e4dd8ad85180cb1ca03ba5e319621f68f2efc2: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5535-3bf96c67884b4da6f3932a8a0e831993fe7037e0-android-x86.apk)
</details>

File: src/core/deltafilewrapper.cpp [hunk 1]
```diff
@@ -641,17 +613,20 @@ void DeltaFileWrapper::addPatch( const QString &localLayerId, const QString &sou
...
+
+    QJsonObject oldFileChecksums;
+    QJsonObject newFileChecksums;
+    addAttachments( localLayerId, tmpOldAttrs, tmpNewAttrs, oldFileChecksums, newFileChecksums );
+    if ( oldFileChecksums.length() > 0 )
+    {
+      oldData.insert( QStringLiteral( "files_sha256" ), oldFileChecksums );
+    }
+
+    if ( newFileChecksums.length() > 0 )
```
Comment (m-kuhn) [inline src/core/deltafilewrapper.cpp:626]:
```suggestion
    if ( !newFileChecksums.isEmpty() )
```

File: src/core/deltafilewrapper.h [hunk 2]
```diff
@@ -378,6 +378,10 @@ class DeltaFileWrapper : public QObject
      */
     bool applyDeltasOnLayers( QHash<QString, QgsVectorLayer *> &vectorLayers, bool shouldApplyInReverse );
 
+    /**
+     * Add file checksums from relevant changed attributes.
+     */
+    void addAttachments( const QString &localLayerId, const QJsonObject &oldAttrs, const QJsonObject &newAttrs, QJsonObject &oldFileChecksums, QJsonObject &newFileChecksums );
```
Comment (m-kuhn) [inline src/core/deltafilewrapper.h:384]:
```suggestion
    /**
     * Add file checksums from relevant changed attributes.
     * \returns a std::tuple<oldChecksums, newChecksums>
     */
    void addAttachments( const QString &localLayerId, const QJsonObject &oldAttrs, const QJsonObject &newAttrs, QJsonObject &oldFileChecksums, QJsonObject &newFileChecksums );
```

We are happy to live in modern days where we no longer have to hack return values through references.

File: src/core/deltafilewrapper.cpp [hunk 3]
```diff
@@ -641,17 +613,20 @@ void DeltaFileWrapper::addPatch( const QString &localLayerId, const QString &sou
   if ( tmpOldAttrs.length() > 0 || tmpNewAttrs.length() > 0 )
```
Comment (m-kuhn) [inline src/core/deltafilewrapper.cpp:613]: could `tmpOldAttrs` be renamed to `changedOldAttrs`? IIUC that's what it refers to

**Comment by m-kuhn** [comment]
Excellent work, thanks.
Is there any chance we can get some accompanying tests in https://github.com/opengisch/QField/blob/master/test/test_deltafilewrapper.cpp ?

File: src/core/deltafilewrapper.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/deltafilewrapper.cpp:613]: @m-kuhn , I've tried to avoid variable renaming noise so the logic tweak is easier to review.

Comment (m-kuhn) [reply src/core/deltafilewrapper.cpp:613]: Good choice.
If it makes sense for you in principle, this can also be done in a follow up PR.

Comment (nirvn) [reply src/core/deltafilewrapper.cpp]: @m-kuhn , yep, that's fair.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**